src/Nethermind/Nethermind.Core/Collections/ArrayPoolExtensions.cs

```diff
@@ -0,0 +1,75 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers;
+
+namespace Nethermind.Core.Collections;
+
+public static class ArrayPoolExtensions
+{
+    /// <summary>
+    /// Rent an array exactly of the same size. May return unpooled array if ArrayPool can't return exact size.
+    /// Always use ReturnExact, or it may throw exception. Don't use this for large array that is probably
+    /// not power of two size.
+    /// </summary>
+    /// <param name="pool"></param>
+    /// <param name="length"></param>
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public static T[] RentExact<T>(this ArrayPool<T> pool, int length)
```

Comment:
I don't like the whole idea of `RentExact`, what you probably want to do is use `ArraySegment` struct or `Memory` probably with custom comparators for the Dictionaries.

The idea here is that most of those Rents will be length misses, which makes it kind of pointless.

Reply:
No, I think most of the time, the key is a keccak, which is 32 byte. It is likely to match.